crates/cli-support/src/js/closures.rs

```diff
@@ -0,0 +1,347 @@
+                fnty.return_type() == Some(ValueType::I32)
+        })
+        .unwrap()
+        .0;
+
+    // And finally, the last piece of the magic. For all our descriptors we
+    // found we inject a JS shim for the descriptor. This JS shim will
+    // manufacture a JS `function`. and prepare it to be invoked.
+    //
+    // Once all that's said and done we inject a new import into the wasm module
+    // of our new wrapper, and the `Remap` step above already wrote calls to
+    // this function within the module.
+    for (i, (_new_idx, descriptor)) in code_idx_to_descriptor {
+        let import_name = format!("__wbindgen_closure_wrapper{}", i);
+
+        let closure = descriptor.closure().unwrap();
+
+        let (js, _ts, _js_doc) = {
+            let mut builder = Js2Rust::new("", input);
+            if closure.mutable {
+                builder
+                    .prelude("let a = this.a;\n")
+                    .prelude("this.a = 0;\n")
+                    .rust_argument("a")
+                    .finally("this.a = a;\n");
+            } else {
+                builder.rust_argument("this.a");
+            }
+            builder
+                .process(&closure.function)?
+                .finish("function", "this.f")
+        };
+        input.expose_add_heap_object();
+        input.function_table_needed = true;
+        let body = format!(
+            "function(ptr, f) {{
+                let cb = {};
+                cb.f = wasm.__wbg_function_table.get(f);
+                cb.a = ptr;
+                let real = cb.bind(cb);
+                real.original = cb;
+                return addHeapObject(real);
+            }}",
+            js,
+        );
+        input.export(&import_name, &body, None);
+
+        let new_import = ImportEntry::new(
+            "__wbindgen_placeholder__".to_string(),
+            import_name,
+            External::Function(type_idx as u32),
+        );
+        input.module.import_section_mut()
+            .unwrap()
+            .entries_mut()
+            .push(new_import);
+    }
+
+    Ok(())
+}
+
+struct Remap<'a> {
+    code_idx_to_descriptor: &'a BTreeMap<u32, (u32, Descriptor)>,
+    old_num_imports: u32,
+}
+
+impl<'a> Remap<'a> {
+    fn remap_module(&self, module: &mut Module) {
+        for section in module.sections_mut() {
+            match section {
+                Section::Export(e) => self.remap_export_section(e),
+                Section::Element(e) => self.remap_element_section(e),
+                Section::Code(e) => self.remap_code_section(e),
+                Section::Start(i) => { self.remap_idx(i); }
+                Section::Name(n) => self.remap_name_section(n),
+                _ => {}
+            }
+        }
+    }
+
+    fn remap_export_section(&self, section: &mut ExportSection) {
+        for entry in section.entries_mut() {
+            self.remap_export_entry(entry);
+        }
+    }
+
+    fn remap_export_entry(&self, entry: &mut ExportEntry) {
```

Comment:
I think this may be a visitor pattern vs trait thing maybe? I personally prefer this pattern as it's less boilerplate-y and serves mostly the same purpose, but I get first dibs here for writing it :)